### ISSUE TITLE
Question about Quick Benchmark Results

### ISSUE THREAD

**Issue by Steamgjk** [eBay/NuRaft#205]
Hi, I am a bit curious about the latency result in https://github.com/eBay/NuRaft/blob/master/docs/bench_results.md 

The network RTT is about 180 micro seconds. Raft needs two RTTs for one request to be committed (Client->Leader->Follower->Leader->Client). In that way, the median latency should be much larger than 180 micro seconds, but why are they almost the same (187 micro seconds)?

**Comment by greensky00** [comment]
Hi @Steamgjk 

As you can find in the benchmark program, client and leader are running in the same process, to measure the pure Raft  performance:
https://github.com/eBay/NuRaft/blob/f004f4c76018819bbe93ce2e5d18c7d466006ad6/tests/bench/raft_bench.cxx#L303-L304

Hence, there is no network cost between client and leader, and each replication can be done within a single RTT.

**Comment by Steamgjk** [comment]
Hi, @greensky00 
I am wondering whether you have some bench results** with >3 replicas**, because according to my perf test, 
with 3 replicas, the throughput can reach ~33K ops/second
but when it comes to 9 replicas, the throughput drops to ~20K ops/second
I am wondering whether this is a normal result, or I miconfigured something.

**Comment by greensky00** [comment]
Hi @Steamgjk 
We don't publish the result with more replicas, but that is expected behavior as the amount of data that the leader should transfer over the network is proportional to the number of followers.

**Comment by Steamgjk** [comment]
Hi, @greensky00 . Do you think it is bounded by network bandwidth or something else? You know, I am using n1-standard-32 VM as replias, the bandwidth is 32^4 Gbps. I feel that is quite large and should be inefficient. So I think it may be more reasonable to attribute the bottleneck to **CPU**, because replicas need to serialize/deserialize/process more messages when we have more replicas. What do you think is a convincing explanation, CPU or bandwidht, or something else?
https://cloud.google.com/compute/docs/machine-types

**Comment by greensky00** [comment]
@Steamgjk 
You can monitor the CPU usage of the leader during the test, and if it reaches 3200%, you can say it is CPU-bound. But I don't believe it uses that much CPU. Network bandwidth is also not likely the cause, as payload size (256 bytes) is small compared to the TCP window size (64KB by default). With simple math, the total throughput of the leader is 256 * 20K * 8 ~= 40MB only for 9 replicas.

Most likely the bottleneck comes from serialization. As I already mentioned in the other comment (https://github.com/eBay/NuRaft/issues/207#issuecomment-836155195), it is not a random and independent data broadcasting. Replication should be strictly and globally ordered, which means log `X` can't be sent to a follower without sending logs up to `X-1`. Since there are more packets for "sending logs up to `X-1`" with more followers, those increased packets are likely waiting for more time in the network queue, and consequently, "sending log `X`" should wait for more time as well.

**Comment by Steamgjk** [comment]
> Most likely the bottleneck comes from serialization. As I already mentioned in the other comment ([#207 (comment)](https://github.com/eBay/NuRaft/issues/207#issuecomment-836155195)), it is not a random and independent data broadcasting. Replication should be strictly and globally ordered, which means log `X` can't be sent to a follower without sending logs up to `X-1`. Since there are more packets for "sending logs up to `X-1`" with more followers, those increased packets are likely waiting for more time in the network queue, and consequently, "sending log `X`" should wait for more time as well.

I agree with it. Actually, one of my lab mates also mentioned this to me. Serialization/deserialization may be the bottleneck
